### ISSUE TITLE
Bug: Not displaying the Submission page after submitting (Trigger problem)

### ISSUE THREAD

**Issue by nmadruga** [openmsupply/application-manager-web-app#1143]
Only happens sometimes -- when the application is re-queried before Actions finish and update the status.

**Comment by CarlosNZ** [comment]
Just to summarise what the problem is here:

The issue is with the way the `useTriggers` is working. The first time `useLoadApplication` runs, it waits to the triggers via `useTriggers` as intended. So this is when loading an application page, or creating a new one. The problem comes when we submit. The mutation causes the data from `useGetApplicationQuery` (in `useLoadApplication`) to refresh, and it does so regardless of whether or not the `useGetApplicationQuery` actually queries the database or not -- i.e. it must update from cache. Therefore, our check of waiting till triggers are ready before re-running the query does not help. The "data" registers as a change, triggering a re-render of all descendant components. So the "ApplicatonSubmission" page gets a "new" version of "structure" that still has DRAFT status and re-directs to the summary page.

I've tried several things to try and make it wait for the triggers, with no success. There doesn't seem to be a way to "intercept" the re-run of the `useGetApplicationQuery` and make it wait for triggers.In theory, we might be able to do something with [onQueryUpdated](https://www.apollographql.com/docs/react/data/refetching/#refetch-options) from the mutation but I've not figured that out.

For now, I think we can do a hacky quick fix -- just make the re-direct to the submission page an HTML `<a>` link rather than a React push, and that should force a full page reload which will wait for triggers as intended. However, this is a very unsatisfactory approach and we need to do something better long-term.

Some possible approaches:

- don't put the `useTriggers` check in `useLoadApplication`. Wait for them both in "ApplicationWrapper" and only render the child routes once both are satisified. However, this may have impacts for the existing `applicationReload` function, which relies on forcing a trigger refetch which then causes `useLoadApplication` to run, so will have to make sure that is taken care of somehow.
- go back to how we used to do it and query the triggers as part of the "useGetApplicationQuery", and keep re-polling until triggers are clear. The reason we went away from this approach though, is because there are cases where we need to check triggers that are not just on the immediately queried table (e.g. when querying a review, we need to also check the triggers on the application table). So if we go with this, we'll need to make sure the queries drill down and fetch all related triggers at the same time, which might be tricky in some cases (which is why we made the `check-triggers` endoint, to move the logic out of the queries).
- more radical re-factor but I think ultimately better: Create a Context for all application data, so application data is always available as global state, which will avoid passing around "structure" as a prop. This is a pretty massive undertaking though, and will have lots of effects, so is not to be undertaken lightly. Ultimately though, I think this is where we want to head (why did we not do this to start with?)

Bear in mind that whatever we do to deal with this will also have to be done for Reviews, as reloading the review after submission will have the same problem.

Anyway, I'll try my hacky solution for now, hopefully that'll get us sorted for the immediate future. Let's not close this issue though, cos it's a significant problem we need to solve.
